Telescope Map : Display if telescope in Night (Astronomical)

Theodlz opened the pull request:
## This PR aims to change the color of the markers of telescope on the telescope map depending on one of the element of their current condition : Day or Night.

### Backend:
For this, we used 2 of the methods defined in the Telescope model: 

- `next_twilight_evening_astronomical`
- `next_twilight_morning_astronomical`

Logically, if the `next_twilight_morning_astronomical` is before the `next_twilight_evening_astronomical`, then it means we are at night (astronomical), and vice versa.

So we added a boolean `is_night_astronomical` to a telescope, so we can use it in the frontend to change the color of the marker depending on that parameter.

### Frontend:

We choose yellow (like the color of the sun) for the color of markers of telescope that are not at night, and a deep purple (like the night sky) for the color of markers of telescope that are at night.

#### Details:
When it comes to nested telescopes (nested telescopes are telescopes that are so close to each other on the map that we grouped them into one marker to avoid overlapping), we show the color associated to that parameter being true if at least one of them has it set to `true`. So a user looking at the map knows that at that location (on this marker), there is at least one of them that is in at night (astronomical).


## Preview:
![telescope_is_night](https://user-images.githubusercontent.com/49785117/166495917-4ade115c-38eb-46a6-9128-124822517903.gif)

mcoughlin commented:
@Theodlz this is cool. I also wonder if also adding information on next twilight / dusk in the text info might combine to give the information you need?

Theodlz commented:
Hi @mcoughlin , do you mean not just saying if its night or day but putting for example the UTC times for the bext twilight and dusk on the list where the telescope info shows up ?

mcoughlin commented:
@Theodlz Yeah or something like that. To give just a bit more info.

Theodlz commented:
Ok ! I'll add it tomorrow then. Have a nice day Michael.

guynir42 reviewed (commented):
The code looks good, and so does the new map page. 
When loading it myself I get this in the console: `Warning: Each child in a list should have a unique "key" prop.` 
Maybe in some instances the component gets an empty list? 

Also, I was wondering if SWIFT is no longer on the map because it doesn't have a fixed location? If so, should there be some way to load the data for such telescopes?

mcoughlin commented:
@guynir42 Yeah I asked @Theodlz to not show the satellites because they didn't make much sense on the world map. Maybe @Theodlz has an idea for a specific dropdown for these?

Theodlz commented:
Hi @guynir42 , @mcoughlin . Yes we didnt show them on the map as their location wouldnt make much sense. As of right now, i can think of anything else then maybe putting a marker (list for the rest of the telescopes) in a specific place on the map (like bottom/top right corner ?), with maybe a bigger size and/or different shape (square not circle) to differenciate it from the rest. So that a user can click that marker and see the list of telescopes with a non fixed location. What do you think ? Easier to implement then yet another drop down list, here we can re use the existing list.

guynir42 commented:
@Theodlz yeah I think a square at the bottom of the map somewhere would be fine. 
Choose a coordinate that puts them somewhere in the ocean where there aren't any other telescopes.

Theodlz commented:
@guynir42 what do you think of it now ?

guynir42 commented:
Looks a lot better. If you can get those non-fixed telescope to not show longitude/latitude that would be even better...

Theodlz commented:
> Looks a lot better. If you can get those non-fixed telescope to not show longitude/latitude that would be even better...

How does it work for this btw, is there a way to get their longitude and latitude if they are changing continuously ? I actually don't know how this works.

mcoughlin commented:
@Theodlz we don't store orbit info, so we won't be able to render anything unfortunately.

Theodlz commented:
> @Theodlz we don't store orbit info, so we won't be able to render anything unfortunately.

Noted. Would be something pretty cool to implement though. I hid the lon lat of non fixed telescopes as you asked @guynir42

guynir42 commented:
> > Looks a lot better. If you can get those non-fixed telescope to not show longitude/latitude that would be even better...
> 
> How does it work for this btw, is there a way to get their longitude and latitude if they are changing continuously ? I actually don't know how this works.

In principle you could probably query some website like "heavens above" to get their current position, but that seems fairly complicated this...

guynir42 commented:
@Theodlz Looks good. I'm thinking either getting rid of the "elevation" as well, or at least filling it in for SWIFT with the actual mean orbital height or something.

Theodlz commented:
> @Theodlz Looks good. I'm thinking either getting rid of the "elevation" as well, or at least filling it in for SWIFT with the actual mean orbital height or something.

Hi @guynir42 , thank your for your feedback. Maybe we should, simply hide it here, and add that feature as a seperate PR ? Otherwise, if we add orbital height, should it just be stored in the elevation ? If so, i can just add it in the demo data maybe ? But i'm not sure that just the orbital height is enough to be useful...

Theodlz commented:
> @Theodlz I think it is really ok to just take off the elevation for space telescopes. It is a cute feature to have that, but no really informative for the user.

Done !

guynir42 commented:
I know I'm being super picky... but can you also remove the sunrise/sunset on the space telescopes, and make the color different maybe? In particular making it yellow when in daytime doesn't make much sense.

Theodlz commented:
> I know I'm being super picky... but can you also remove the sunrise/sunset on the space telescopes, and make the color different maybe? In particular making it yellow when in daytime doesn't make much sense.

Like this @guynir42 ?

![telescopes_non_fixed](https://user-images.githubusercontent.com/49785117/167796162-ecee0223-869b-42ec-8044-a0f9cb9469a9.gif)

Theodlz commented:
As we noticed with @mcoughlin , some of the lon and lat in the demo data were wrong. One more thing i want to add before merging (important), is to avoid the page from crashing if it receives telescopes with wrong lon lat (out of the range it expects for the angles). So it can be a bit more robust.

Theodlz commented:
> @Theodlz I've already merged a PR with a fix to the wrong longitudes. No need to do anything else. Please merge master and check that everything works.

Amazing ! Just saw it. So should i remove the verification i added in the handler ?

guynir42 commented:
@Theodlz I think you can remove your verification if it looks like it works with out it (after pulling in the latest). 
As for the demo data, I think we can leave it as it is, if the only "problem" is that some of the longitudes are bigger than 180. That's not really problematic, it only causes problems with the timezone finder.

Theodlz commented:
> @Theodlz I think you can remove your verification if it looks like it works with out it (after pulling in the latest). As for the demo data, I think we can leave it as it is, if the only "problem" is that some of the longitudes are bigger than 180. That's not really problematic, it only causes problems with the timezone finder.

Great @guynir42. I left the frontend verification for lon lat as a wrong longitude could potentially cause issues with displaying the map. But i removed the modifications i added to the backend and demo data.

Theodlz commented:
Ok @guynir42, i ended up removing verification i added yesterday from both frontend and backend. Seems that react-simple-maps handles these 'out of range' angles pretty well.

guynir42 reviewed (approved):
LGTM

mcoughlin reviewed (approved):
Thanks for all the iteration on this @Theodlz
